marshall + size with vtproto

TheSafo opened the pull request:
### What does this PR do?

### Motivation

### Describe how you validated your changes

### Additional Notes

joyzhang-dd committed:
Initial implementation of stateful gRPC stream for Logs agent
Hooks gRPC stream to Logs agent pipeline

joyzhang-dd committed:
Handle state management and snapshot transmission

TheSafo committed:
move stateful encoding proto to proper location
adapted the stateful transport implementation to import the new location

joyzhang-dd committed:
Implement backoff strategy at stream level covering common failures

TheSafo committed:
modify proto

joyzhang-dd committed:
Seperate out stream Send into its own goroutine

During flow-control/back-pressure, a slow/blocking Send won't block the supervisor loop

DDuongNguyen committed:
latest logs agent pattern and intake changes

TheSafo committed:
update protos again with tags and package

TheSafo committed:
Adding compression for stateful encoding transport

TheSafo committed:
fix gomod

DDuongNguyen committed:
implement tag manager + refactor

TheSafo committed:
Add initial metrics

joyzhang-dd committed:
Change pattern ID from UUID to monotonic increasing ID

TheSafo committed:
gomod tidy via dda inv -e create-module pkg/logs/patterns/tags, dda inv modules.add-all-replace dda inv -e tidy

DDuongNguyen committed:
implement pattern eviction, fix tests, lint, add go mod

TheSafo committed:
marshall + size with vtproto

nathan-b commented on pkg/logs/patterns/tags/tag_manager.go:
```diff
@@ -0,0 +1,140 @@
+
+// Get returns the dictionary ID for a tag string, if it exists
+// Returns the ID and a boolean indicating if the tag was found
+// Deprecated: Use GetStringID instead
```
If this is new functionality is there really any value in checking in a deprecated method?

nathan-b reviewed (approved):
I think the only reason I'm on this review is because of the go mod tidy. As this is going into your private branch and the code doesn't touch ARUN files I'll approve even though this PR is a bit of a mess 😅

joyzhang-dd committed:
Stateful-encoding transport optimization

It contains the following:
1. simplified pattern updates (no delete+define, now define=update)
2. encodes a wildcard value with type inference (i.e. use uint or float when possible)
3. single dictionary entry for all tags, encoded as tagset (contains proto change)

joyzhang-dd committed:
Implement delta-encoding for certains fields in stateful-encoding transport

Note the current implementation delta-encode at batch-level, which doesn't require
protobuf change. We still make the protobuf change in case we want to extend
the delta-encoding at stream-level in the future

DDuongNguyen committed:
implement JSON preprocessing for structured logs and add json_context to protobuf

TheSafo committed:
delta encoding timestamps as int64s instead of uint64

DDuongNguyen committed:
implement tag eviction, refactor eviction logic into smaller packages, small fixes on some tests, add configs for tag eviction logic, go mod cleaning

TheSafo committed:
remove size() metrics

TheSafo committed:
fix compiling

TheSafo committed:
go mod tidy
